File: src/ui/js/Backups.js
/**
 *  Eternity Keeper, a Pillars of Eternity save game editor.
 *  Copyright (C) 2016 the authors.
 *
 *  Eternity Keeper is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  Eternity Keeper is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// File -> Backups: the copies the editor takes before it deletes, renames or
// replaces a save in the saves folder, and a way to put one back. Restoring
// never replaces a save that is there (save/SaveBackups refuses), so the only
// thing this page has to get right is saying what each copy is.
var Backups = function () {
	var self = this;

	var defaultState = {
		loading: false
		, folder: ''
		, keep: 10
		, backups: []
		, restored: {}      // backup id -> true, once put back this session
		, restoring: null   // the id being put back
		, message: ''
		, failed: false
	};

	self.state = $.extend({}, defaultState);
	self.html = {};

	self.init = () => {
		self.html.menuBackups.click(() => self.open());
		self.html.backupsOpenFolder.click(() => self.ask({action: 'open'}, () => {}));
	};

	self.render = newState => {
		self.state = $.extend({}, defaultState, newState);
		var state = self.state;

		self.html.backupsIntro.text('Before Eternity Keeper deletes, renames or replaces a save, '
			+ 'it keeps a copy here. The ' + state.keep + ' most recent are kept'
			+ (state.folder ? ', in ' + state.folder + '.' : '.'));

		var list = self.html.backupsList.empty();
		if (state.loading) {
			list.append($('<p class="bk-empty">').text('Reading the backups…'));
		} else if (state.backups.length < 1) {
			list.append($('<p class="bk-empty">').text('No backups yet. One appears here the '
				+ 'first time the editor deletes, renames or replaces a save.'));
		} else {
			state.backups.forEach(backup => list.append(self.row(backup)));
		}

		self.html.backupsStatus.text(state.message)
			.toggleClass('bk-failed', state.failed)
			.toggle(!!state.message);
	};
};

Backups.prototype.open = function () {
	this.transition({loading: true, message: '', failed: false, restored: {}});
	this.html.backupsDialog.modal('show');
	this.refresh();
};

Backups.prototype.refresh = function () {
	var self = this;
	self.ask({action: 'list'}, reply => self.transition({
		loading: false
		, folder: reply.folder
		, keep: reply.keep
		, backups: reply.backups
	}));
};

Backups.prototype.ask = function (request, then) {
	var self = this;
	window.backups({
		request: JSON.stringify(request)
		, onSuccess: response => then(JSON.parse(response))
		, onFailure: (code, message) => self.transition({
			loading: false, restoring: null, message: message, failed: true})
	});
};

Backups.prototype.restore = function (backup) {
	var self = this;
	if (self.state.restoring) {
		return;
	}

	self.transition({restoring: backup.id, message: '', failed: false});
	self.ask({action: 'restore', id: backup.id}, reply => {
		var restored = $.extend({}, self.state.restored);
		restored[backup.id] = true;

		// The save list only shows what it last read, so read it again -- but
		// only while it is on screen: searching switches to the list, and that
		// would pull the user out of a save they have open.
		var onList = !!(Eternity.state && Eternity.state.listView);
		if (onList) {
			Eternity.SaveSearch.search();
		}

		self.transition({
			restoring: null
			, restored: restored
			, failed: false
			, message: 'Put back as ' + reply.restored + '.'
				+ (onList ? '' : ' It will be on the save list the next time you search.')
		});
	});
};

Backups.prototype.row = function (backup) {
	var self = this;
	var name = backup.userSaveName || backup.sceneTitle || backup.file;
	var details = [
		backup.sceneTitle && backup.sceneTitle !== name ? backup.sceneTitle : ''
		, 'copied ' + backup.why
		, self.when(backup.time)
		, (backup.size / 1048576).toFixed(1) + ' MB'
	].filter(part => part);

	var action;
	if (self.state.restored[backup.id]) {
		action = $('<span class="bk-done">').text('Restored');
	} else {
		action = $('<button type="button" class="btn btn-primary btn-sm bk-restore">')
			.text(self.state.restoring === backup.id ? 'Restoring…' : 'Restore')
			.prop('disabled', !!self.state.restoring)
			.attr('title', 'Put this copy back as ' + backup.original)
			.click(() => self.restore(backup));
	}

	return $('<div class="bk-row">')
		.append($('<div class="bk-what">')
			.append($('<div class="bk-name">').text(name))
			.append($('<div class="bk-details">').text(details.join(' · ')))
			.append($('<div class="bk-file">').text(backup.file)))
		.append(action);
};

Backups.prototype.when = function (time) {
	var date = new Date(time);
	return date.toLocaleDateString(undefined, {day: 'numeric', month: 'short', year: 'numeric'})
		+ ', ' + date.toLocaleTimeString(undefined, {hour: '2-digit', minute: '2-digit'});
};

$.extend(Backups.prototype, Renderer.prototype);

File: src/main/java/uk/me/mantas/eternity/handlers/Backups.java
/**
 *  Eternity Keeper, a Pillars of Eternity save game editor.
 *  Copyright (C) 2016 the authors.
 *
 *  Eternity Keeper is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  Eternity Keeper is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package uk.me.mantas.eternity.handlers;

import org.cef.browser.CefBrowser;
import org.cef.callback.CefQueryCallback;
import org.cef.handler.CefMessageRouterHandlerAdapter;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import uk.me.mantas.eternity.Logger;
import uk.me.mantas.eternity.environment.Environment;
import uk.me.mantas.eternity.save.SaveBackups;
import uk.me.mantas.eternity.save.SaveBackups.Backup;

import java.awt.Desktop;
import java.io.File;
import java.io.IOException;

/**
 * The Backups dialog's side of {@link SaveBackups}:
 * {@code {"action": "list"}} answers every copy with where it came from and
 * why; {@code {"action": "restore", "id"}} puts one back in its folder, never
 * over a save that is there; {@code {"action": "open"}} shows the folder.
 */
public class Backups extends CefMessageRouterHandlerAdapter {
	private static final Logger logger = Logger.getLogger(Backups.class);

	@Override
	public boolean onQuery (
		final CefBrowser browser
		, final long id
		, final String request
		, final boolean persistent
		, final CefQueryCallback callback) {

		// A restore copies a whole save; not on the browser's thread.
		Environment.getInstance().workers().execute(() -> answer(request, callback));
		return true;
	}

	private void answer (final String request, final CefQueryCallback callback) {
		final JSONObject json;
		try {
			json = new JSONObject(request);
		} catch (final JSONException e) {
			callback.failure(-1, "Unreadable request: " + e.getMessage());
			return;
		}

		final SaveBackups backups = SaveBackups.forThisProcess();
		final String action = json.optString("action", "list");
		try {
			switch (action) {
				case "list":
					callback.success(list(backups).toString());
					break;

				case "restore":
					final File restored = backups.restore(json.getString("id"));
					callback.success(new JSONObject()
						.put("restored", restored.getAbsolutePath()).toString());
					break;

				case "open":
					open(backups.folder());
					callback.success(new JSONObject().put("opened", true).toString());
					break;

				default:
					callback.failure(-1, "Unknown backups action: " + action);
			}
		} catch (final IOException | JSONException | UnsupportedOperationException e) {
			logger.error("Backups %s failed: %s%n", action, e.getMessage());
			callback.failure(-1, e.getMessage());
		}
	}

	private static JSONObject list (final SaveBackups backups) {
		final JSONArray rows = new JSONArray();
		for (final Backup backup : backups.list()) {
			rows.put(new JSONObject()
				.put("id", backup.id)
				.put("file", backup.file.getName())
				.put("original", backup.original.getAbsolutePath())
				.put("reason", backup.reason.name())
				.put("why", backup.reason.description)
				.put("time", backup.time)
				.put("size", backup.size)
				.put("userSaveName", backup.userSaveName)
				.put("sceneTitle", backup.sceneTitle));
		}

		return new JSONObject()
			.put("folder", backups.folder().getAbsolutePath())
			.put("keep", SaveBackups.KEEP)
			.put("backups", rows);
	}

	private static void open (final File folder) throws IOException {
		if (!folder.isDirectory() && !folder.mkdirs()) {
			throw new IOException("Could not create " + folder.getAbsolutePath() + ".");
		}

		if (!Desktop.isDesktopSupported()) {
			throw new UnsupportedOperationException(
				"This system cannot open folders from here. The backups are in "
					+ folder.getAbsolutePath() + ".");
		}

		Desktop.getDesktop().open(folder);
	}

	@Override
	public void onQueryCanceled (final CefBrowser browser, final long id) {
		logger.error("Query #%d was cancelled.%n", id);
	}
}
